From b80705b93d51af2c06bcde48deec8296370a1fa4 Mon Sep 17 00:00:00 2001
From: Leo Yu-Chi Liang <ycliang@andestech.com>
Date: Mon, 28 Oct 2024 10:59:42 +0800
Subject: [PATCH 165/172] andes: fix Smepmp setting for ATCSMU mmio range
 (#123)

Makatau series CPU supports Smepmp extension, so we need to add
PMP setting for MMIO region that may be accessed in M-mode (ATCSMU).
However, we have already used up 8 PMP entries, which is also the maximum number
of PMP entry on AX45MP-8C bitmap and no more PMP entry we can use, therefore
we first check if the bitmap is Makatau serires CPU before using another PMP entry.

After this patch (on AX65)
Domain0 Region00 : 0x00000000f0100000-0x00000000f0100fff M: (I,R,W) S/U: (R,W)	// ATCSMU
Domain0 Region01 : 0x00000000f0300000-0x00000000f0300fff M: (I,R,W) S/U: (R,W)	// UART
Domain0 Region02 : 0x00000000e0500000-0x00000000e050ffff M: (I,R,W) S/U: (R,W)	// L2 cache
Domain0 Region03 : 0x0000000000000000-0x000000000003ffff M: (R,X) S/U: ()	// opensbi R/X region
Domain0 Region04 : 0x0000000000040000-0x000000000007ffff M: (R,W) S/U: ()	// opensbi R/W region
Domain0 Region05 : 0x00000000e6000000-0x00000000e60fffff M: (I,R,W) S/U: ()	// PLIC
Domain0 Region06 : 0x00000000e6400000-0x00000000e67fffff M: (I,R,W) S/U: ()	// PLIC-SW
Domain0 Region07 : 0x00000000e4000000-0x00000000e5ffffff M: (I,R,W) S/U: (R,W)	// PLMT
Domain0 Region08 : 0x0000000000000000-0xffffffffffffffff M: () S/U: (R,W,X)	// Else

After this patch (on AX45MP-8C)
Domain0 Region00 : 0x00000000f0300000-0x00000000f0300fff M: (I,R,W) S/U: (R,W)	// UART
Domain0 Region01 : 0x00000000e0500000-0x00000000e050ffff M: (I,R,W) S/U: (R,W)	// L2 cache
Domain0 Region02 : 0x0000000000000000-0x000000000003ffff M: (R,X) S/U: ()	// opensbi R/X region
Domain0 Region03 : 0x0000000000040000-0x000000000007ffff M: (R,W) S/U: ()	// opensbi R/W region
Domain0 Region04 : 0x00000000e6000000-0x00000000e60fffff M: (I,R,W) S/U: ()	// PLIC
Domain0 Region05 : 0x00000000e6400000-0x00000000e67fffff M: (I,R,W) S/U: ()	// PLIC-SW
Domain0 Region06 : 0x00000000e4000000-0x00000000e5ffffff M: (I,R,W) S/U: (R,W)	// PLMT
Domain0 Region07 : 0x0000000000000000-0xffffffffffffffff M: () S/U: (R,W,X)	// Else

Fixes: Bugzilla 32040 (http://e-andes.andestech.com/bugzilla5/show_bug.cgi?id=32040)
Signed-off-by: Leo Yu-Chi Liang <ycliang@andestech.com>
Reviewed-on: https://gitea.andestech.com/RD-SW/opensbi/pulls/123
Reviewed-by: Mina Hui-Min Chou <minachou@andestech.com>
Reviewed-by: Tim Shih-Ting OuYang <tim609@andestech.com>
Reviewed-by: Charles Ci-Jyun Wu <dminus@andestech.com>
---
 platform/generic/andes/ae350.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/platform/generic/andes/ae350.c b/platform/generic/andes/ae350.c
index fec6248939a3..9158ffc14adf 100644
--- a/platform/generic/andes/ae350.c
+++ b/platform/generic/andes/ae350.c
@@ -333,6 +333,11 @@ static void ae350_smu_device_init(void)
 		sbi_hsm_set_device(&andes_smu_hsm);
 		sbi_system_suspend_set_device(&andes_smu_susp);
 
+		if ((csr_read(CSR_MARCHID) & 0xff) == 0x65)
+			rc = sbi_domain_root_add_memrange(smu.addr, 0x1000, 0x1000,
+							  SBI_DOMAIN_MEMREGION_MMIO |
+							  SBI_DOMAIN_MEMREGION_SHARED_SURW_MRW);
+
 		/* Allocate space for regs that need to be saved/restored */
 		save_regs_off = sbi_scratch_alloc_offset(sizeof(struct save_regs));
 		if (!save_regs_off)
@@ -346,7 +351,6 @@ static int ae350_final_init(bool cold_boot, const struct fdt_match *match)
 		return 0;
 
 	pma_init();
-	ae350_smu_device_init();
 	trigger_init();
 
 	return 0;
@@ -354,8 +358,10 @@ static int ae350_final_init(bool cold_boot, const struct fdt_match *match)
 
 static int ae350_early_init(bool cold_boot, const struct fdt_match *match)
 {
-	if (cold_boot)
+	if (cold_boot) {
+		ae350_smu_device_init();
 		return fdt_cache_init();
+	}
 
 	if (save_regs_off)
 		ae350_suspend_non_ret_restore(sbi_scratch_thishart_ptr(), false);
-- 
2.34.1

